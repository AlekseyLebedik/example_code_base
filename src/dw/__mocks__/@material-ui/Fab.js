import Fab from '@material-ui/core/Fab';

jest.mock('@material-ui/core/Fab', () => 'Fab');

export default Fab;
