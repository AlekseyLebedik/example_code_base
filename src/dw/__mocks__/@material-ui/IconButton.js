import IconButton from '@material-ui/core/IconButton';

jest.mock('@material-ui/core/IconButton', () => 'IconButton');

export default IconButton;
