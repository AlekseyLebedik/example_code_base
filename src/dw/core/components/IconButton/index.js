import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Fab from '@material-ui/core/Fab';
import Icon from '@material-ui/core/Icon';
import IconButton from '@material-ui/core/IconButton';
import Tooltip from '@material-ui/core/Tooltip';
import Zoom from '@material-ui/core/Zoom';

const styles = theme => ({
  root: {
    transition: '0.3s',
  },
  colorPrimary: {
    color: theme.palette.action.active,
    '&:hover': {
      color: theme.palette.primary.main,
    },
  },
  colorSecondary: {
    color: theme.palette.action.active,
    '&:hover': {
      color: theme.palette.secondary.main,
    },
  },
});

const displayIcon = ({
  classes,
  className,
  color,
  dataCy,
  disabled,
  icon,
  iconProps,
  onClick,
  size,
  text,
  type,
  variant,
}) => {
  switch (type) {
    case 'fab':
      return (
        <Fab
          className={className}
          color={color}
          data-cy={dataCy}
          disabled={disabled}
          onClick={onClick}
          variant={variant}
        >
          <Icon fontSize={size} {...iconProps}>
            {icon}
          </Icon>
          {text}
        </Fab>
      );
    case 'button':
      return (
        <Button
          className={className}
          color={color}
          data-cy={dataCy}
          disabled={disabled}
          onClick={onClick}
          size="small"
          variant={variant}
        >
          <Icon fontSize={size} {...iconProps}>
            {icon}
          </Icon>
        </Button>
      );
    default:
      return (
        <IconButton
          classes={{
            root: `${classes.root} ${className && className}`,
            colorPrimary: classes.colorPrimary,
            colorSecondary: classes.colorSecondary,
          }}
          color={color}
          data-cy={dataCy}
          disabled={disabled}
          onClick={onClick}
          onKeyDown={e => {
            if (e.key === 'Enter') {
              e.preventDefault();
              e.target.querySelector('span.material-icons').click();
            }
          }}
          type="button"
        >
          <Icon fontSize={size} {...iconProps}>
            {icon}
          </Icon>
        </IconButton>
      );
  }
};

displayIcon.propTypes = {
  classes: PropTypes.object.isRequired,
  className: PropTypes.string,
  dataCy: PropTypes.string,
  color: PropTypes.oneOf(['inherit', 'default', 'primary', 'secondary']),
  disabled: PropTypes.bool,
  icon: PropTypes.string.isRequired,
  iconProps: PropTypes.object,
  onClick: PropTypes.func,
  size: PropTypes.oneOf(['inherit', 'default', 'small', 'large']),
  text: PropTypes.string,
  tooltip: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  type: PropTypes.oneOf(['icon', 'fab', 'button']),
  variant: PropTypes.oneOf(['text', 'outlined', 'contained', 'extended']),
};

displayIcon.defaultProps = {
  className: '',
  color: 'default',
  dataCy: null,
  disabled: false,
  iconProps: {},
  onClick: () => {},
  size: 'default',
  text: '',
  type: 'icon',
  variant: 'text',
};

const WrappedIcon = ({ tooltip, ...props }) => (
  <Tooltip title={tooltip} TransitionComponent={Zoom}>
    {props.disabled ? <span>{displayIcon(props)}</span> : displayIcon(props)}
  </Tooltip>
);

WrappedIcon.propTypes = displayIcon.propTypes;

export default withStyles(styles)(WrappedIcon);
