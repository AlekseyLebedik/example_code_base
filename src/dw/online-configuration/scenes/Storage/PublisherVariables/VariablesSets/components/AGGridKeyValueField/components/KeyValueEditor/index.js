import React, { useCallback, useMemo, useRef, useState } from 'react';
import PropTypes from 'prop-types';

import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Fab from '@material-ui/core/Fab';
import IconButton from '@material-ui/core/IconButton';
import Icon from '@material-ui/core/Icon';
import Tooltip from '@material-ui/core/Tooltip';
import TextField from '@material-ui/core/TextField';
import Dialog from 'dw/core/components/Dialog';

const useStyles = makeStyles(theme => ({
  addButton: {
    margin: `0 0 ${theme.spacing(1)}px ${theme.spacing(2)}px`,
    width: 43,
    zIndex: 1,
  },
}));

const KeyValueEditor = ({ gridApi, node, onDataChanged }) => {
  const valueRef = useRef();
  const classes = useStyles();
  const [open, setOpen] = useState(false);
  const [key, setKey] = useState(node?.data?.key || '');
  const [value, setValue] = useState(node?.data?.value || '');
  const [error, setError] = useState();

  const onClose = useCallback(() => {
    setKey(node?.data?.key || '');
    setValue(node?.data?.value || '');
    setError(null);
    setOpen(false);
  }, [node]);

  const onSave = useCallback(() => {
    setError(null);
    const existing = [];
    gridApi.forEachLeafNode(({ data }) => existing.push(data.key));
    if (!key) {
      setError('Required');
      return;
    }
    if (existing.includes(key) && key !== node?.data?.key) {
      setError('The key already exists');
      return;
    }
    if (node) {
      node.setData({ key, value });
    } else {
      gridApi.applyTransaction({
        addIndex: 0,
        add: [{ key, value }],
      });
    }
    onDataChanged({ api: gridApi });
    if (!node) {
      setKey('');
      setValue('');
    }
    setOpen(false);
    setTimeout(() => {
      // Highlight the changed row
      gridApi.flashCells({
        rowNodes: [node || gridApi.getDisplayedRowAtIndex(0)],
        cellFlashDelay: 2000,
      });
    }, 0);
  }, [gridApi, node, key, value]);

  const actions = useMemo(
    () => [
      <Button onClick={onClose}>Cancel</Button>,
      <Button onClick={onSave} color="primary">
        Save
      </Button>,
    ],
    [onSave]
  );
  const handleKeyPress = useCallback(
    event => {
      if (event.key === 'Enter') {
        if (event.target.name === 'key') valueRef.current.focus();
        else onSave();
      }
    },
    [onSave]
  );

  return (
    <>
      {node ? (
        <Tooltip title="Edit key / value pair" placement="left">
          <IconButton onClick={() => setOpen(true)}>
            <Icon>edit</Icon>
          </IconButton>
        </Tooltip>
      ) : (
        <Tooltip title="Add a new key / value pair" placement="left">
          <Fab
            color="primary"
            size="small"
            onClick={() => setOpen(true)}
            className={classes.addButton}
          >
            <Icon>playlist_add</Icon>
          </Fab>
        </Tooltip>
      )}
      {open && (
        <Dialog
          title={node ? 'Edit Key / Value pair' : 'Add new Key / Value pair'}
          actions={actions}
          onRequestClose={onClose}
          fullWidth
          open
          modal
        >
          <div className="flex flex-col">
            <TextField
              name="key"
              value={key}
              label="Key"
              onChange={e => setKey(e.target.value)}
              error={Boolean(error)}
              helperText={error}
              InputLabelProps={{ shrink: true }}
              onKeyPress={handleKeyPress}
              fullWidth
              autoFocus
            />
            <TextField
              name="value"
              inputRef={valueRef}
              value={value}
              label="Value"
              onChange={e => setValue(e.target.value)}
              InputLabelProps={{ shrink: true }}
              onKeyPress={handleKeyPress}
              fullWidth
            />
          </div>
        </Dialog>
      )}
    </>
  );
};
KeyValueEditor.propTypes = {
  gridApi: PropTypes.object.isRequired,
  node: PropTypes.object,
  onDataChanged: PropTypes.func.isRequired,
};
KeyValueEditor.defaultProps = {
  node: undefined,
};

export default KeyValueEditor;
