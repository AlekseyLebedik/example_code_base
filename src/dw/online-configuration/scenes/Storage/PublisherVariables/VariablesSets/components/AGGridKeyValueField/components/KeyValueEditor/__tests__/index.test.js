import React from 'react';

import { shallow } from 'enzyme';

import IconButton from '@material-ui/core/IconButton';
import Fab from '@material-ui/core/Fab';

import KeyValueEditor from '..';

const gridApi = {};
const onDataChanged = jest.fn();

describe('KeyValueEditor', () => {
  it('render add new key / value', () => {
    const wrapper = shallow(
      <KeyValueEditor gridApi={gridApi} onDataChanged={onDataChanged} />
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.find(Fab).simulate('click');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
  it('render edit key / value', () => {
    const wrapper = shallow(
      <KeyValueEditor
        gridApi={gridApi}
        onDataChanged={onDataChanged}
        node={{ data: { key: 1, value: 1 } }}
      />
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.find(IconButton).simulate('click');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
