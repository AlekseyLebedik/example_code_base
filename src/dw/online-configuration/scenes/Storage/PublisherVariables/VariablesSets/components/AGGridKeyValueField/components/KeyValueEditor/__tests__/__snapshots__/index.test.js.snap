// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyValueEditor render add new key / value 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Tooltip))
    placement="left"
    title="Add a new key / value pair"
  >
    <WithStyles(ForwardRef(Fab))
      className="makeStyles-addButton-1"
      color="primary"
      onClick={[Function]}
      size="small"
    >
      <WithStyles(ForwardRef(Icon))>
        playlist_add
      </WithStyles(ForwardRef(Icon))>
    </WithStyles(ForwardRef(Fab))>
  </WithStyles(ForwardRef(Tooltip))>
</Fragment>
`;

exports[`KeyValueEditor render add new key / value 2`] = `
<Fragment>
  <WithStyles(ForwardRef(Tooltip))
    placement="left"
    title="Add a new key / value pair"
  >
    <WithStyles(ForwardRef(Fab))
      className="makeStyles-addButton-1"
      color="primary"
      onClick={[Function]}
      size="small"
    >
      <WithStyles(ForwardRef(Icon))>
        playlist_add
      </WithStyles(ForwardRef(Icon))>
    </WithStyles(ForwardRef(Fab))>
  </WithStyles(ForwardRef(Tooltip))>
  <DialogComponent
    actions={
      Array [
        <WithStyles(ForwardRef(Button))
          onClick={[Function]}
        >
          Cancel
        </WithStyles(ForwardRef(Button))>,
        <WithStyles(ForwardRef(Button))
          color="primary"
          onClick={[Function]}
        >
          Save
        </WithStyles(ForwardRef(Button))>,
      ]
    }
    cancelOnBackdropClick={false}
    classes={Object {}}
    contentStyle={Object {}}
    fullWidth={true}
    maxWidth="sm"
    modal={true}
    onRequestClose={[Function]}
    open={true}
    title="Add new Key / Value pair"
  >
    <div
      className="flex flex-col"
    >
      <WithStyles(ForwardRef(TextField))
        InputLabelProps={
          Object {
            "shrink": true,
          }
        }
        autoFocus={true}
        error={false}
        fullWidth={true}
        label="Key"
        name="key"
        onChange={[Function]}
        onKeyPress={[Function]}
        value=""
      />
      <WithStyles(ForwardRef(TextField))
        InputLabelProps={
          Object {
            "shrink": true,
          }
        }
        fullWidth={true}
        inputRef={
          Object {
            "current": undefined,
          }
        }
        label="Value"
        name="value"
        onChange={[Function]}
        onKeyPress={[Function]}
        value=""
      />
    </div>
  </DialogComponent>
</Fragment>
`;

exports[`KeyValueEditor render edit key / value 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Tooltip))
    placement="left"
    title="Edit key / value pair"
  >
    <WithStyles(ForwardRef(IconButton))
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(Icon))>
        edit
      </WithStyles(ForwardRef(Icon))>
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(Tooltip))>
</Fragment>
`;

exports[`KeyValueEditor render edit key / value 2`] = `
<Fragment>
  <WithStyles(ForwardRef(Tooltip))
    placement="left"
    title="Edit key / value pair"
  >
    <WithStyles(ForwardRef(IconButton))
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(Icon))>
        edit
      </WithStyles(ForwardRef(Icon))>
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(Tooltip))>
  <DialogComponent
    actions={
      Array [
        <WithStyles(ForwardRef(Button))
          onClick={[Function]}
        >
          Cancel
        </WithStyles(ForwardRef(Button))>,
        <WithStyles(ForwardRef(Button))
          color="primary"
          onClick={[Function]}
        >
          Save
        </WithStyles(ForwardRef(Button))>,
      ]
    }
    cancelOnBackdropClick={false}
    classes={Object {}}
    contentStyle={Object {}}
    fullWidth={true}
    maxWidth="sm"
    modal={true}
    onRequestClose={[Function]}
    open={true}
    title="Edit Key / Value pair"
  >
    <div
      className="flex flex-col"
    >
      <WithStyles(ForwardRef(TextField))
        InputLabelProps={
          Object {
            "shrink": true,
          }
        }
        autoFocus={true}
        error={false}
        fullWidth={true}
        label="Key"
        name="key"
        onChange={[Function]}
        onKeyPress={[Function]}
        value={1}
      />
      <WithStyles(ForwardRef(TextField))
        InputLabelProps={
          Object {
            "shrink": true,
          }
        }
        fullWidth={true}
        inputRef={
          Object {
            "current": undefined,
          }
        }
        label="Value"
        name="value"
        onChange={[Function]}
        onKeyPress={[Function]}
        value={1}
      />
    </div>
  </DialogComponent>
</Fragment>
`;
